### PR TITLE
docs: Improve Ingress documentation

### PR DESCRIPTION
**Pull request by youngnick**
This PR changes the formatting of the Service Mesh documentation
slightly to make the new sections added more discoverable.

It also adds a section clarifying Cilium's path types and precedence
when using Ingress, along with an example that illustrates this more
clearly. 

Lastly, it also splits out the Ingress and Network Policy tutorials into their
own specific example, and add an explanation at the top level of how
the `reserved:ingress` identity works. As part of this, some existing doc
fragments have been converted into reusable pieces.

### CODE REVIEW

**Review by mhofstetter — APPROVED** [review]
Thanks for this update!

File: Documentation/network/servicemesh/ingress.rst [hunk 1]
```diff
@@ -61,8 +61,105 @@ Prerequisites
...
+  ``/foo/bar`` contains no regex characters, when it is configured in Envoy as a
+  regex, it will function as an ``Exact`` match instead).
+
+When multiple path types are configured on an Ingress object, Cilium will configure
+Envoy with the matches in the following order:
+
+* Exact
+* ImplementationSpecific (that is, regular expression)
+* Prefix
+* The ``/`` Prefix match has special handling and always goes last.
```
Comment (ghost) [inline Documentation/network/servicemesh/ingress.rst]:
```suggestion
#. Exact
#. ImplementationSpecific (that is, regular expression)
#. Prefix
#. The ``/`` Prefix match has special handling and always goes last.
```
nit, but in this case it makes sense to use an ordered list

File: Documentation/network/servicemesh/ingress.rst [hunk 2]
```diff
@@ -61,8 +61,105 @@ Prerequisites
...
+If you do use ImplementationSpecific regex support, be careful with using the
+``*`` operator, since it will increase the length of the regex, but may match
+another, shorter option.
+
+For example, if you have two ImplementationSpecific paths, ``/impl``, and ``/impl.*``,
+the second will be sorted ahead of the first in the generated config. But because
+``*`` is in use, the ``/impl`` match will never be hit, as any request to that
+path will match the ``/impl.*`` path first.
+
+See the Ingress Path Types Example below for more information.
```
Comment (ghost) [inline Documentation/network/servicemesh/ingress.rst]: I'd rather link the relevant page/section here instead of saying "below".

**Review by ghost — APPROVED** [review]
I left two non-blocking comments. The structure looks good to me, thanks!

**Review by jrajahalme — APPROVED** [review]
Checked the policy related parts! IMO this is a great improvement on the docs :-)

**Review by nathanjsweet — APPROVED** [review]

File: Documentation/network/servicemesh/ingress.rst (hunk 2 above)
Comment (sayboras) [reply Documentation/network/servicemesh/ingress.rst]: Good point and thanks :+1:

**Pull request merged**